### PR TITLE
[AUS] expose metrics

### PR DESCRIPTION
**Pull request by geoberle**
Advanced Upgrade Service is not a regular qontract-reconcile integration but rather an SRE capability with a different support model. AUS will not fail as a mean to signal an issue if it detects something problematic during reconciling. Instead it will expose enough metrics so that fine grained alerts can be defined.

The `aus_cluster_upgrade_policy_info` and `aus_cluster_upgrade_policy_info` metrics are going to be used for reporting purposes.

All metrics are exposed by all flavours of cluster upgrade integrations (the addons upgrade scheduler exposes only counters for reconcile loops & failures per org for now).

See https://service.pages.redhat.com/dev-guidelines/docs/appsre/expose-metrics/ for more information about transactional and scoped metric exposure.

## aus_organization_validation_errors

Gauge - Current validation errors within an OCM organization

```
aus_organization_validation_errors{integration="advanced-upgrade-scheduler",ocm_env="ocm-stage",org_id="xxx"} 0.0
```

## aus_organization_reconciled_total

Counter - Counter for the number of times an OCM organization was reconciled

```
aus_organization_reconciled_total{integration="advanced-upgrade-scheduler",ocm_env="ocm-stage",org_id="xxx"} 1.0
```

## aus_organization_reconcile_error_total

Counter - Counter for the failed reconcile runs for an OCM organization

```
aus_organization_reconcile_error_total{integration="advanced-upgrade-scheduler",ocm_env="ocm-stage",org_id="xxx"} 1.0
```

## aus_cluster_upgrade_policy_info

Info - An info metric for cluster upgrade policy details under AUS upgrade control

```
aus_cluster_upgrade_policy_info{cluster_name="my-cluster",cluster_uuid="some-uuid",integration="advanced-upgrade-scheduler",mutexes="",ocm_env="ocm-stage",org_id="xxx",schedule="* * * * *",sector="",soak_days="1",workloads="test"} 1.0
```

## aus_cluster_version_remaining_soak_days

Gauge - Remaining days a version needs to soak for a cluster

```
aus_cluster_version_remaining_soak_days{cluster_uuid="some-uid",integration="advanced-upgrade-scheduler",ocm_env="ocm-stage",soaking_version="4.12.15"} 5.20
```

part of https://issues.redhat.com/browse/APPSRE-7360

### CODE REVIEW

File: reconcile/aus/metrics.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    mutexes: str
+    soak_days: str
+    workloads: str
+
+    @classmethod
+    def name(cls) -> str:
+        return "aus_cluster_upgrade_policy_info"
+
+
+class AUSOrganizationValidationErrorMetric(AUSBaseMetric, GaugeMetric):
```
Comment (patjlm) [inline reconcile/aus/metrics.py:45]: don't you want to name that according to the metric type?  `AUSOrganizationValidationErrorGauge` ? Maybe also add the `s` at `Error` ?

The same applies to `AUSClusterVersionRemainingSoakDaysMetric` -> `AUSClusterVersionRemainingSoakDaysGauge` above

File: reconcile/aus/metrics.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+
+class AUSOrganizationReconcileErrorCounter(AUSBaseMetric, CounterMetric):
+    "Counter for the failed reconcile runs for an OCM organization"
+
+    org_id: str
+
+    @classmethod
+    def name(cls) -> str:
+        return "aus_organization_reconcile_error"
```
Comment (patjlm) [inline reconcile/aus/metrics.py:72]: there is an `s` in `aus_organization_validation_errors` above. Maybe add one here as well?

File: reconcile/aus/ocm_upgrade_scheduler.py [hunk 3]
```diff
@@ -94,3 +109,31 @@ def get_ocm_env_upgrade_specs(
...
+                continue
+
+            workload_soaking_upgrades = [
+                aus.soaking_days(version_data, upgrades, wl, False)
+                for wl in up.workloads
+            ]
+            for version in upgrades or []:
+                soaks = [s.get(version, 0) for s in workload_soaking_upgrades]
+                metrics.set_gauge(
+                    AUSClusterVersionRemainingSoakDaysMetric(
```
Comment (patjlm) [inline reconcile/aus/ocm_upgrade_scheduler.py:134]: Should we use a scoped transactional metric context here in order to automatically remove policies/clusters that disappeared?

Comment (patjlm) [reply reconcile/aus/ocm_upgrade_scheduler.py:134]: also `version` would disappear

Comment (geoberle) [reply reconcile/aus/ocm_upgrade_scheduler.py:134]: all exposed metrics are under a scoped transaction

https://github.com/app-sre/qontract-reconcile/blob/bf6f6e8053ad14a439029a116012a53101823104/reconcile/aus/base.py#L78

File: reconcile/aus/metrics.py (hunk 2 above)
Comment (geoberle) [reply reconcile/aus/metrics.py]: good point. fixed

File: reconcile/aus/metrics.py (hunk 1 above)
Comment (geoberle) [reply reconcile/aus/metrics.py]: good point. fixed

File: reconcile/aus/ocm_upgrade_scheduler.py (hunk 3 above)
Comment (patjlm) [reply reconcile/aus/ocm_upgrade_scheduler.py]: ah did not go up the stack enough! all good!

**Pull request merged**